### PR TITLE
Refactor ancestor lookup

### PR DESCRIPTION
**Pull request by terencechain**
From audit feedback, the ancestor lookup should fail over from fork choice store to db. Given such feedback, I also refactored from `ancestor` into  `ancestorByForkChoiceStore` and `ancestorByDB` for readability. I added comments for clarify and a regression test. Tested run time.

### CODE REVIEW

File: beacon-chain/blockchain/process_block_helpers.go [hunk 1]
```diff
@@ -258,16 +258,35 @@ func (s *Service) ancestor(ctx context.Context, root []byte, slot uint64) ([]byt
 	ctx, span := trace.StartSpan(ctx, "forkChoice.ancestor")
 	defer span.End()
 
-	// Stop recursive ancestry lookup if context is cancelled.
-	if ctx.Err() != nil {
-		return nil, ctx.Err()
-	}
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_block_helpers.go:264]: Why remove?

File: beacon-chain/blockchain/process_block_helpers.go [hunk 2]
```diff
@@ -258,16 +258,35 @@ func (s *Service) ancestor(ctx context.Context, root []byte, slot uint64) ([]byt
...
+
+	return ar, nil
+}
+
+// This retrieves an ancestor root using fork choice store. The look up is looping through the a flat array structure.
+func (s *Service) ancestorByForkChoiceStore(ctx context.Context, r [32]byte, slot uint64) ([]byte, error) {
 	if s.forkChoiceStore.HasParent(r) {
-		return s.forkChoiceStore.AncestorRoot(ctx, r, slot)
+		return nil, errors.New("could not find root in fork choice store")
+	}
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_block_helpers.go:281]: This branch doesn't make sense to me. "If fork choice store has the parent, return an error"? Why?

Comment (terencechain) [reply beacon-chain/blockchain/process_block_helpers.go:281]: This should be `!s.forkChoiceStore.HasParent(r)`

It shows a big test gap. I need to figure out what went wrong and add more tests here

File: beacon-chain/blockchain/process_block_helpers.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/blockchain/process_block_helpers.go:264]: Discussed offline, this check is done in the recursive / expensive methods.

File: beacon-chain/blockchain/process_block_helpers.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block_helpers.go]: Turns out that existing code path was not tested 😱  I just added a test + the fix. Thanks for catching this!

File: beacon-chain/blockchain/process_block_helpers.go [hunk 3]
```diff
@@ -258,16 +258,35 @@ func (s *Service) ancestor(ctx context.Context, root []byte, slot uint64) ([]byt
...
+	return s.forkChoiceStore.AncestorRoot(ctx, r, slot)
+}
+
+// This retrieves an ancestor root using DB. The look up is recursively looking up DB. Slower than `ancestorByForkChoiceStore`.
+func (s *Service) ancestorByDB(ctx context.Context, r [32]byte, slot uint64) ([]byte, error) {
+	// Stop recursive ancestry lookup if context is cancelled.
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
 	}
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_block_helpers.go:290]: Want to add a quick test for this path? Coverage is complaining untested.

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**